db/crud.go

```diff
@@ -896,34 +896,42 @@ func (db *DatabaseCollectionWithUser) OnDemandImportForWrite(ctx context.Context
 }
 
 // updateHLV updates the HLV in the sync data appropriately based on what type of document update event we are encountering
-func (db *DatabaseCollectionWithUser) updateHLV(d *Document, docUpdateEvent DocUpdateType) (*Document, error) {
+func (db *DatabaseCollectionWithUser) updateHLV(d *Document, previousMou *MetadataOnlyUpdate, docUpdateEvent DocUpdateType) (*Document, error) {
 
+	hasHLV := d.HLV != nil
 	if d.HLV == nil {
 		d.HLV = &HybridLogicalVector{}
 	}
 	switch docUpdateEvent {
 	case ExistingVersion:
 		// preserve any other logic on the HLV that has been done by the client, only update to cvCAS will be needed
 		d.HLV.CurrentVersionCAS = expandMacroCASValueUint64
-		d.HLV.ImportCAS = 0 // remove importCAS for non-imports to save space
 	case Import:
-		if d.HLV.CurrentVersionCAS == d.Cas {
-			// if cvCAS = document CAS, the HLV has already been updated for this mutation by another HLV-aware peer.
-			// Set ImportCAS to the previous document CAS, but don't otherwise modify HLV
-			d.HLV.ImportCAS = d.Cas
-		} else {
+		// Update HLV if:
+		//
+		// - local SDK write without existing HLV. There could be a stale _mou, but that should be ignored.)
+		// - local SDK write with an existing HLV, where there was a document already processed by local cluster.
+		// 	- _vv.cvCAS != document cas
+		// 	- no _mou or _mou.cas != document cas
+
+		// Do not update HLV if this is an XDCR migration:
+		//
+		// - document was written by another cluster
+		// 	- _vv.cvCAS != document cas, no _mou or _mou.cas != document cas
+		// - document was written by another cluster, itself having an undergone an import
+		// 	- _vv.cvCAS == document cas and _mou.cas != document cas
+
+		if !hasHLV || (d.HLV.CurrentVersionCAS != d.Cas && (previousMou == nil || base.HexCasToUint64(previousMou.CAS) != d.Cas)) {
```

Comment:
If this can be rewritten to match the comment suggestion I think that would be more readable.  Even something like this maybe:
```
cvCASMatch := hasHLV && d.HLV.CurrentVersionCAS == d.Cas
mouMatch := previousMou != nil && base.HexCasToUint64(previousMou.CAS) == d.Cas)
if !cvCASMatch && !mouMatch {
```

Reply:
There is one additional caveat, which I've express in the changed code, but I'm not sure if it is expressed correctly in the comment. There could be no existing `_vv` but an `_mou`, in which case we are going to write one. I'm not sure this is terribly likely, but I wanted to include it if a document did get corrupted or someone else is using `_mou` and not using `_vv` for some reason.